packages/legend-query-builder/src/stores/explorer/QueryBuilderPropertySearchState.ts

```diff
@@ -41,45 +45,102 @@ import {
 import type { QueryBuilderState } from '../QueryBuilderState.js';
+import { Fuse } from './CJS__Fuse.cjs';
 
 export class QueryBuilderPropertySearchState {
   queryBuilderState: QueryBuilderState;
   // TODO: Check if we could clean this up as this seems quite complicated and its purpose is not clear to me
   // See https://github.com/finos/legend-studio/pull/1212
-  allMappedPropertyNodes: QueryBuilderExplorerTreeNodeData[] = [];
+  mappedPropertyNodes: QueryBuilderExplorerTreeNodeData[] = [];
   searchedMappedPropertyNodes: QueryBuilderExplorerTreeNodeData[] = [];
+
   isSearchPanelOpen = false;
+  isSearchPanelTipsOpen = false;
   isSearchPanelHidden = false;
+  isOverSearchLimit = false;
+
   searchText = '';
+  searchState = ActionState.create().pass();
+  searchEngine: Fuse<QueryBuilderExplorerTreeNodeData>;
+
   filterByMultiple: boolean;
+
   typeFilters: QUERY_BUILDER_PROPERTY_SEARCH_TYPE[];
 
+  modeOfSearch: SEARCH_MODE;
+  modeOfSearchOptions: SEARCH_MODE[];
+
   constructor(queryBuilderState: QueryBuilderState) {
     makeAutoObservable(this, {
       queryBuilderState: false,
       searchedMappedPropertyNodes: observable,
       isSearchPanelOpen: observable,
+      isSearchPanelTipsOpen: observable,
+      modeOfSearch: observable,
+      isOverSearchLimit: observable,
       isSearchPanelHidden: observable,
       searchText: observable,
+      searchEngine: observable,
       filteredPropertyNodes: computed,
       setSearchText: action,
+      setIsOverSearchLimit: action,
       setSearchedMappedPropertyNodes: action,
       setIsSearchPanelOpen: action,
+      setIsSearchPanelTipsOpen: action,
       setIsSearchPanelHidden: action,
-      refreshPropertyState: action,
+      resetPropertyState: action,
       setFilterByMultiple: action,
       toggleTypeFilter: action,
+      changeModeOfSearch: action,
     });
 
     this.queryBuilderState = queryBuilderState;
     this.filterByMultiple = true;
+
+    this.modeOfSearch = SEARCH_MODE.NORMAL;
+
+    this.modeOfSearchOptions = [
+      SEARCH_MODE.NORMAL,
+      SEARCH_MODE.INCLUDE,
+      SEARCH_MODE.EXACT,
+      SEARCH_MODE.INVERSE,
+    ];
+
     this.typeFilters = [
       QUERY_BUILDER_PROPERTY_SEARCH_TYPE.CLASS,
       QUERY_BUILDER_PROPERTY_SEARCH_TYPE.STRING,
       QUERY_BUILDER_PROPERTY_SEARCH_TYPE.BOOLEAN,
       QUERY_BUILDER_PROPERTY_SEARCH_TYPE.NUMBER,
       QUERY_BUILDER_PROPERTY_SEARCH_TYPE.DATE,
     ];
+
+    this.searchEngine = new Fuse(this.mappedPropertyNodes);
+  }
+
+  setFuse(): void {
+    this.searchEngine = new Fuse(this.mappedPropertyNodes, {
+      includeScore: true,
+      shouldSort: true,
+      minMatchCharLength: QUERY_BUILDER_PROPERTY_SEARCH_TEXT_MIN_LENGTH,
+      ignoreLocation: true,
+      threshold: 0.3,
+      keys: [
+        {
+          name: 'id',
+        },
+        {
+          name: 'label',
+        },
+        {
+          name: 'taggedValues',
+          getFn: (node: QueryBuilderExplorerTreeNodeData) =>
```

Comment:
didn't we agree to only scan by the `doc.doc` `tagged value`? See `CORE_PURE_PATH.PROFILE_DOC`

Reply:
why do we need this cast? we could from the beginning restricted this to `QueryBuilderExplorerTreePropertyNodeData` right?

Reply:
I think perhaps because `this.mappedproperties` type itself is `QueryBuilderExplorerTreeNodeData[]` which can contain multiple things so program will complain 
"Argument of type 'QueryBuilderExplorerTreeNodeData[]' is not assignable to parameter of type 'readonly QueryBuilderExplorerTreePropertyNodeData[]'.
  Type 'QueryBuilderExplorerTreeNodeData' is missing the following properties from type 'QueryBuilderExplorerTreePropertyNodeData': property, parentId" without the cast 🤔